autonat.go

```diff
@@ -0,0 +1,160 @@
+package autonat
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	host "github.com/libp2p/go-libp2p-host"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type NATStatus int
+
+const (
+	NATStatusUnknown NATStatus = iota
+	NATStatusPublic
+	NATStatusPrivate
+)
+
+var (
+	AutoNATBootDelay       = 15 * time.Second
+	AutoNATRefreshInterval = 15 * time.Minute
+
+	AutoNATRequestTimeout = 60 * time.Second
+)
+
+type AutoNAT interface {
+	Status() NATStatus
+	PublicAddr() (ma.Multiaddr, error)
+}
+
+type AmbientAutoNAT struct {
+	ctx    context.Context
+	host   host.Host
+	peers  map[peer.ID]struct{}
+	status NATStatus
+	addr   ma.Multiaddr
+	mx     sync.Mutex
```

Comment:
We started using anonymous/nested struct pattern to describe what is protected by mutex.
I am not 100% sure if the `peers` map is the only thing but if it is it would look:

```go
peers struct {
    sync.Mutex
    set map[peer.ID]struct{}
}
```

Reply:
See also: https://talks.golang.org/2012/10things.slide#3

Reply:
the mutex protect the state/address as well.

Reply:
mutex hat in https://github.com/libp2p/go-libp2p-autonat/pull/1/commits/cd7a8757fa205a98942214585d08fa977e430c11